### PR TITLE
Prefer native parser for SSH public key parsing

### LINKED ISSUE #23363: Docker/rootless: "ssh-keygen": executable file not found in $PATH -
### Description

```
Cannot verify your SSH key: SSHKeyGenParsePublicKey: fail to parse public key: exec: "ssh-keygen": executable file not found in $PATH -
```

After clicking on "Add deploy key" with the form in the screenshot

On `https://git.domain.tld/<org>/<repo>/settings/keys`

That said I tried to verify my SSH key that I had added to my profile and it worked fine, so maybe it uses another code function.

### Gitea Version

1.19.0-rc1

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

image

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

With Docker/docker-compose

```yml
version: "2.3"

services:
    my_git:
        image: gitea/gitea:1.19.0-rc1-rootless
```

I am pretty sure it works fine with `gitea/gitea:1.19.0-rc1`

### Database

MySQL

### PR DESCRIPTION
**Pull request by yogo1212**
Without this patch, the setting SSH.StartBuiltinServer decides whether the native (Go) implementation is used rather than calling 'ssh-keygen'. It's possible for 'using ssh-keygen' and 'using the built-in server' to be independent.
In fact, the gitea rootless container doesn't ship ssh-keygen and can be configured to use the host's SSH server - which will cause the public key parsing mechanism to break.

This commit changes the decision to be based on SSH.KeygenPath instead. Any existing configurations with a custom KeygenPath set will continue to function.  The new default value of '' selects the native version. The downside of this approach is that anyone who has relying on plain 'ssh-keygen' to have special properties will now be using the native version instead.
I assume the exec-variant is only there because /x/crypto/ssh didn't support ssh-ed25519 until 2016. I don't see any other reason for using it so it might be an acceptable risk.

Fixes #23363

EDIT: this message was garbled when I tried to get the commit description back in.. Trying to reconstruct it:

## :warning: BREAKING :warning: Users who don't have SSH.KeygenPath explicitly set and rely on the ssh-keygen binary need to set SSH.KeygenPath to 'ssh-keygen' in order to be able to continue using it for public key parsing.

There was something else but I can't remember at the moment.

EDIT2: It was about `make test` and `make lint`. Can't get them to run.
To reproduce the issue, I installed `golang` in `docker.io/node:16` and got:
```
...
go: mvdan.cc/xurls/v2@v2.4.0: unknown revision mvdan.cc/xurls/v2.4.0
go: gotest.tools/v3@v3.4.0: unknown revision gotest.tools/v3.4.0
...
go: gotest.tools/v3@v3.0.3: unknown revision gotest.tools/v3.0.3
...
go: error loading module requirements
```

### CODE REVIEW

**Review by williamdes — COMMENTED** [review]
Thanks

**Review by lunny — APPROVED** [review]

File: docs/content/doc/administration/config-cheat-sheet.en-us.md [hunk 1]
```diff
@@ -343,7 +343,7 @@ The following configuration set `Content-Type: application/vnd.android.package-a
 - `SSH_SERVER_MACS`: **hmac-sha2-256-etm@openssh.com, hmac-sha2-256, hmac-sha1**: For the built-in SSH server, choose the MACs to support for SSH connections, for system SSH this setting has no effect
 - `SSH_SERVER_HOST_KEYS`: **ssh/gitea.rsa, ssh/gogs.rsa**: For the built-in SSH server, choose the keypairs to offer as the host key. The private key should be at `SSH_SERVER_HOST_KEY` and the public `SSH_SERVER_HOST_KEY.pub`. Relative paths are made absolute relative to the `APP_DATA_PATH`. If no key exists a 4096 bit RSA key will be created for you.
 - `SSH_KEY_TEST_PATH`: **/tmp**: Directory to create temporary files in when testing public keys using ssh-keygen, default is the system temporary directory.
-- `SSH_KEYGEN_PATH`: **ssh-keygen**: Path to ssh-keygen, default is 'ssh-keygen' which means the shell is responsible for finding out which one to call.
+- `SSH_KEYGEN_PATH`: **ssh-keygen**: Use ssh-keygen to parse public SSH keys. The value is passed to the shell. By default, gitea does the parsing itself.
```
Comment (wxiaoguang) [inline docs/content/doc/administration/config-cheat-sheet.en-us.md]:
```suggestion
- `SSH_KEYGEN_PATH`: **\<empty\>**: Use `ssh-keygen` to parse public SSH keys. The value is passed to the shell. By default, Gitea does the parsing itself.
```

**Comment by wxiaoguang** [comment]
If it doesn't change the behavior for daily usage, or it doesn't affect end users, I guess it's not that breaking? @yardenshoham

**Comment by lunny** [comment]
> If it doesn't change the behavior for daily usage, or it doesn't affect end users, I guess it's not that breaking? @yardenshoham

I would also like to add that break hint so people could know what changed when view release notes. Of course, it should not affect anything.

**Comment by yogo1212** [comment]
The option was added to deal with selinux limiting access to `/tmp` ([issue](https://github.com/gogs/gogs/issues/2179), [PR](https://github.com/gogs/gogs/pull/2637)) ("and even adjust the ssh-keygen path").
Why didn't they just send the key via STDIN?! (`ssh-keygen -lf -`)
`KeygenPath` should never have been added, imho. Hard to take it back now..
Ironically, the merger of ed25519 support in `/x/crypto/ssh` and the mentioned PR are just two months apart.
("At the end it should be possible to use the go native ssh library only without any dependency on ssh-keygen at all" they said while they added the option that makes doing that a breaking change. Some behaviour is best left unspecified)

The breaking scenario is: 1. it was necessary to have a special version of `ssh-keygen` AND 2. that replacement was placed on the `PATH` env of gitea - without modifying `KeygenPath`.
Those two users with magic versions for `ssh-keygen` are hopefully using the `KeygenPath`.

I chose this approach to get a solution that breaks as little as possible without increasing the complexity.
My initial approach was to just add `openssh-keygen` in the rootless container. Other users should be able to just install the package on their machine. But tacking on more is not always the best solution.
(hmm. forgot to change branch name... my official motto for this PR is now: "I hope I didn't forget anything important")

**Comment by silverwind** [comment]
How does go ssh compare openssh? is it kept up to date? I do consider OpenSSH the reference implementation, so my gut reaction is that this is a step back.

**Comment by yogo1212** [comment]
@silverwind this change is only concerned with parsing the public key in order to learn its length and the key type. the key formats are specified in RFCs and there really shouldn't be much difference between how [go ssh](https://pkg.go.dev/golang.org/x/crypto/ssh#ParsePublicKey) and openssh treat them.

i would say it's of little concern.
those who are concerned can set  `KeygenPath` to `ssh-keygen`.

**Comment by silverwind** [comment]
Should we maybe, as a first step, only change the rootless container to use builtin ssh-keygen, and leave the other setups at the openssh version?

**Review by silverwind — APPROVED** [review]
Well, let's try it then.

**Comment by lunny** [comment]
CI failure is related.

**Comment by yogo1212** [comment]
@lunny @jolheiser The failing test calls `SSHKeyGenParsePublicKey` directly. Taat fails because `SSHKeyGenParsePublicKey` expects KeygenPath to be set and the default for `KeygenPath` is now empty.
The function isn't normally called when `KeygenPath` is empty.

I've added two commits:
One adding an new test for the native parser.
Another to allow the (unconfigured) test to use the keygen-based parser by having a default inside the function.

I've decided against passing in the keygen path to `SSHKeyGenParsePublicKey` from the outside. A simple solution would be to add a parameter but then the signature now longer matches that of the native version.
Another solution would've been to change `settings` before calling the test but that also doesn't seem nice.

**Comment by yogo1212** [comment]
@lunny @silverwind thank you for your work on gitea and nice days to you both :-)

**Pull request merged**